### PR TITLE
Documentation on how to upgrade Agones and/or Kubernetes.

### LINKED ISSUE #555: Document how to do upgrades of Agones
Right now the upgrade experience is likely quite tricky -- and risky.

Best bet right now may be to advocate spinning up a new cluster, and moving to a new cluster at upgrade time, and if so, we should document this.

RBAC rule changes particularly frighten me for upgrades.

We should test out some scenarios and see where things fall over, and document them (likely also find some issues a long the way).

### PR DESCRIPTION
**Pull request by markmandel**
Documentation is deliberately the safest upgrade paths.

Closes #555

### CODE REVIEW

**Comment by markmandel** [comment]
Preview of the page: https://7226a9c-dot-preview-dot-agones-images.appspot.com/site/docs/installation/upgrading/

File: site/content/en/docs/Installation/upgrading.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+version of Kubernetes.  This means that we can slowly migrate our game servers from the older version's cluster (blue) 
+to the new version's cluster (green), and ensure nothing surprising happens during this process.
+
+This also allows easy rollback to the previous infrastructure that we already know to be working in production,
+with minimal interruptions to player experience.
+
+The following are steps to implement this:
+
+1. Create a new cluster of the same size or smaller as the current cluster, with the new version of Kubernetes
+2. Install the same version of Agones on the new cluster, as you have on the previous cluster.
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/upgrading.md]: This is useful if you want to separate the upgrade of k8s vs agones. But I imagine most folks will actually want to do them together (for minor upgrades) since agones will only be stress tested against a single patch version of k8s at a time.

File: site/content/en/docs/Installation/upgrading.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+7. Congratulations - you have now upgraded to a new version of Kubernetes! 👍
+
+### Single Cluster
+
+If you are upgrading a single cluster, we recommend creating a maintenance window, in which your game goes offline
+for the period of your upgrade, as there will be a short period in which Agones will be non-responsive during the node
+upgrades.
+
+1. Start your maintenance window.
+1. Scale your Fleets down to 0 and/or delete your GameServers.
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/upgrading.md]: can you explain why? will node upgrades kill allocated game servers? is it possible to set PDBs and/or graceful termination periods on them automatically to make node upgrades work without disruption?

File: site/content/en/docs/Installation/upgrading.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+1. Start and complete your node upgrades.
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/upgrading.md]: The k8s docs recommend doing master upgrades before node upgrades.

File: site/content/en/docs/Installation/upgrading.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+1. Start and complete you master upgrade.
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/upgrading.md]: upgrade(s) -- since folks, especially on prem or when not using a provider, will likely have a multi-master setup.

File: site/content/en/docs/Installation/upgrading.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+The following are strategies for safely upgrading Agones from one version to another. They may require adjustment to 
+your particular game architecture but should provide a solid foundation for updating Agones safely.
+
+The recommended approach is to use [multiple clusters](#multiple-clusters), such that the upgrade can be tested
+gradually with production load and easily rolled back if need arises.
+
+### Multiple Clusters
+
+Essentially, we want to do a [blue/green deployment](https://martinfowler.com/bliki/BlueGreenDeployment.html) to the
```
[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Lemme see if I can reword it. The process is essentially the same, but slower (:man_shrugging:) but if it doesn't work, maybe we should cut it.

[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Actually, changed my mind. Cutting it entirely, and rewording.

File: site/content/en/docs/Installation/upgrading.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+5. Assuming everything is working successfully on the new cluster, slowly increase the percentage of matched sessions to the new cluster, until you reach 100%.
+6. Once you are comfortable with the stability of the new cluster with the new Agones version, shut down the old cluster.
+7. Congratulations - you have now upgraded to a new version of Agones! 👍
+
+### Single Cluster
+
+If you are upgrading a single cluster, we recommend creating a maintenance window, in which your game goes offline
+for the period of your upgrade, as there will be a short period in which Agones will be non-responsive during the upgrade.
+
+### Installation with install.yaml
```
[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Whoops. Good catch.

File: site/content/en/docs/Installation/upgrading.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+1. Run any other tests to ensure the Agones installation is working as expected.
+1. Close your maintenance window.
+7. Congratulations - you have now upgraded to a new version of Agones! 👍
+
+### Installation with Helm
+
+Helm features capabilities for upgrading to newer versions without having to delete the older version. For details on
+how to use Helm for upgrades, see the [helm upgrade](https://helm.sh/docs/helm/#helm-upgrade) documentation.
+
+Given the above, the steps for upgrade are simpler:
```
[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: I'd _really_ like to break apart and reorganise the installation docs as they currently stand -- they are one big mung of things, that diverge only sometimes for random topics.

I started to dig into what you have above, because I think we should do it, but it just made me want to pull this whole thing apart more - as then i want a section on "why choose which installation method", but then I was just unravelling more and more string as I pulled on the thread.

TL;DR - can we do this later?

File: site/content/en/docs/Installation/upgrading.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+1. Start your maintenance window.
+1. Run `helm upgrade` with the appropriate arguments, such a `--version`, for your specific upgrade
+1. Run any other tests to ensure the Agones installation is working as expected.
+1. Close your maintenance window.
+7. Congratulations - you have now upgraded to a new version of Agones! 👍
+
+## Upgrading Kubernetes
```
[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: So I'm not sure what the right text here is. I like what you have here re: risk on minor vs patch upgrades, but I wonder if the text about doing them "in place" is the best advice. My concern is that if you do it in place, you have to have a maintenance window, because you can't do with live gameservers running. Even for patch upgrades, doing it with two clusters means you can do it while your game is active, without downtime.

So I'm not sure what the right thing to say here is :man_shrugging: Maybe just the section on differences in risk between versions? Especially if you have custom code that also interacts with the K8s api?

File: site/content/en/docs/Installation/upgrading.md (hunk 1 above)
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Added a note to this effect.  PTAL. I don't _think_ we need a combined K8s + Agones upgrade at the same time doc? Or do you think we do?

File: site/content/en/docs/Installation/upgrading.md (hunk 3 above)
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Ah good catch, switching.

I went looking for a URL that called this out in the K8s docs, but couldn't fine one. Do you have one at hand? Would be good to link to it.

File: site/content/en/docs/Installation/upgrading.md (hunk 2 above)
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Added some words to the docs - but the TL;DR is basically to prevent #398 .

We can mitigate this with separate nodepools for the controller, but for this guide, I figured it was best to write the simplest version I could, with strategies I knew would work for all circumstances. I figured we could then iterate out to more sophisticated upgrade paths as needed. 

The other option is to split this into a "separate controller node group" section, and a "everything on the same node group" section, with seperate upgrade strategies, but wasn't sure if that was too much for this instance.

I err'd towards a path that I knew would work for all use cases, since I figured stable was best for 1.0.

Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: > is it possible to set PDBs and/or graceful termination periods on them automatically to make node upgrades work without disruption

On this note - it's _possible_ (probably more graceful termination, we don't have control over which Pods get terminated with PDBs AFAIK?), but I don't think it's been thoroughly tested as a strategy, so I didn't want to document it at this stage. Just trying to be safe, essentially.

File: site/content/en/docs/Installation/upgrading.md (hunk 1 above)
Comment (roberthbailey) [reply site/content/en/docs/Installation/upgrading.md]: I agree that we don't need another section (or doc) on doing them together. Just a note that as long as you are upgrading agones and either switching clusters or having maintenance that you might as well make sure that you are running the right k8s version underneath it.

File: site/content/en/docs/Installation/upgrading.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (roberthbailey) [reply site/content/en/docs/Installation/upgrading.md]: This one is tricky. I'm not sure if it's reasonable to force new cluster creation and workload migration each time there is a security patch (which seem relatively frequent). But I see your point about maintenance windows if you do in-place.... Leaving it as-is seems reasonable for now, and we can work on some of the trickier edge cases in a follow up.

File: site/content/en/docs/Installation/upgrading.md (hunk 2 above)
Comment (roberthbailey) [reply site/content/en/docs/Installation/upgrading.md]: That sounds reasonable. I agree that we can expand on the instructions over time.

File: site/content/en/docs/Installation/upgrading.md (hunk 3 above)
Comment (roberthbailey) [reply site/content/en/docs/Installation/upgrading.md]: It took me a while looking, and the only place I could find it was buried in the version skew documentation: https://kubernetes.io/docs/setup/release/version-skew-policy/#kubelet

I know that having the master be newer than the nodes is enforced in GKE and kubeadm (and likely elsewhere) fwiw.

File: site/content/en/docs/Installation/upgrading.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+We essentially want to transition our GameServer allocations from a cluster with the old version of Kubernetes,
+to a cluster with the upgraded version of Kubernetes while ensuring nothing surprising 
+happens during this process.
+
+This also allows easy rollback to the previous infrastructure that we already know to be working in production, with
+minimal interruptions to player experience.
+
+
+Essentially, we want to do a [blue/green deployment](https://martinfowler.com/bliki/BlueGreenDeployment.html) to the new
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/upgrading.md]: Did you mean to leave this part in here, after removing it from the above section?

File: site/content/en/docs/Installation/upgrading.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+The following are strategies for safely upgrading the underlying Kubernetes cluster from one version to another.
+They may require adjustment to your particular game architecture but should provide a solid foundation for updating your cluster safely.
+
+The recommended approach is to use [multiple clusters](#multiple-clusters-1), such that the upgrade can be tested
+gradually with production load and easily rolled back if need arises.
+
+Agones currently has only a [single supported Kubernetes version]({{< relref "_index.md#usage-requirements" >}}) for each version,
+so it is recommended to do supported minor (1.12.1 ➡ 1.13.2) Kubernetes version upgrades at the same time as matching Agones upgrades.
+
+Patch upgrades (e.g. 1.12.1 ➡ 1.12.3) within the same major version of Kubernetes can be done at any time.
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/upgrading.md]: s/major/minor

File: site/content/en/docs/Installation/upgrading.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+## Upgrading Kubernetes
+
+The following are strategies for safely upgrading the underlying Kubernetes cluster from one version to another.
+They may require adjustment to your particular game architecture but should provide a solid foundation for updating your cluster safely.
+
+The recommended approach is to use [multiple clusters](#multiple-clusters-1), such that the upgrade can be tested
+gradually with production load and easily rolled back if need arises.
+
+Agones currently has only a [single supported Kubernetes version]({{< relref "_index.md#usage-requirements" >}}) for each version,
+so it is recommended to do supported minor (1.12.1 ➡ 1.13.2) Kubernetes version upgrades at the same time as matching Agones upgrades.
```
Comment (roberthbailey) [inline site/content/en/docs/Installation/upgrading.md]: i'm not sure how i feel about this arrow... it's a bit distracting in the rendered paragraph. Feel free to leave it in if you like the way it looks. :)

File: site/content/en/docs/Installation/upgrading.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Agreed on tricky. Also, depending on your Kubernetes control layer (cloud, kubeadm, etc), there are likely different techniques you can deploy. e.g. nodepools are GKE specific, but maybe with a combo of regional cluster + multple nodepools, you can do a complete upgrade in cluster, within minimal/practically no downtime?

File: site/content/en/docs/Installation/upgrading.md (hunk 3 above)
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Cool. I'll not worry about a link. We can have the same order in our docs too :+1:

File: site/content/en/docs/Installation/upgrading.md (hunk 9 above)
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Doh. Thanks.

File: site/content/en/docs/Installation/upgrading.md (hunk 11 above)
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: -> feels so boring compared to :arrow_right: 

:smile:

File: site/content/en/docs/Installation/upgrading.md (hunk 1 above)
Comment (markmandel) [reply site/content/en/docs/Installation/upgrading.md]: Done!

> Agones currently has only a [single supported Kubernetes version]({{< relref "_index.md#usage-requirements" >}}) for each version,
so it is recommended to do supported minor (e.g. 1.12.1 ➡ 1.13.2) Kubernetes version upgrades at the same time as matching Agones upgrades.
